Drag and drop in new search page

PenghaiZhang opened the pull request:
#688 

##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [x] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This PR adds the feature of 'Drag and Drop' to the new Search UI so each Item and attachment is draggable and they can be dropped to the course list in Selection Session.

The approach is basically to reuse some old JS functions provided in `courselist.js` to make TS component draggable. 

The whole logic should be: 

1.  `CourseListSection.java` prepares a droppable area which is the course list, including setting up the `onDrop` event handler, and specifying what elements can be dropped.

2. A search is in done in new search UI and returns some search result Items.

3. Make each search result Item draggable.

4. After the async call for attachment viewer details is completed, make each attachment draggable.


Please view this [Video](https://streamable.com/768cpf).

edalex-ian commented:
Heya @PenghaiZhang 

Just noticed in the video we're still missing the 'drag indicator' in front of item titles. Any chance you could add?

PenghaiZhang commented:
> Heya @PenghaiZhang
> 
> Just noticed in the video we're still missing the 'drag indicator' in front of item titles. Any chance you could add?

Sorry I totally forgot. Added and updated the video.

edalex-ian commented on Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:
```diff
@@ -703,3 +706,26 @@ describe("conversion of parameters to SearchPageOptions", () => {
+    await renderSearchPage();
+
+    getSearchResult.results.forEach(({ uuid }) => {
+      expect(
```
There's a small chance here that this test could pass even though no search results have been marked draggable. This is because the expect block is in a function that will only be called if there are 'results'.

Now obviously we're sure that results has entries, but.... Nowhere here have we confirmed that, so if that mock gets messed up somewhere else we'll never know. Maybe it'd be valid to have an expect above line 725 which confirms `getSearchResult.results` has a size greater than zero. :thinking: 

Up to you.

edalex-ian commented on Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/components/SearchResult.test.tsx:
```diff
@@ -306,5 +310,15 @@ describe("<SearchResult/>", () => {
+      updateMockGetRenderData(basicRenderData);
+      await renderSearchResult(mockData.attachSearchObj);
+      mockData.attachSearchObj.attachments!.forEach((attachment) => {
+        expect(
```
Similar concern as above with potential for `expect` never being called.

edalex-ian commented on Source/Plugins/Core/com.equella.core/resources/web/scripts/courselist.js:
```diff
@@ -63,115 +63,148 @@ var CourseList = {
 		var itemResultSelector = '.selectable.' + itemResultClass;
 		var itemSummarySelector = "h2." + itemSummaryClass;
 		var attachmentSelector = '.selectable .attachmentrow.with-select';
+		var newSearchPageItemSelector = '.NewSearchPage-Item';
+		var newSearchPageAttachmentSelector = '.NewSearchPage-Attachment';
```
This is a super minor thing - so feel free to ignore - but, do we really need to prefix the classes (which end up in the DOM and on the React side) with `New`. Could these just be:

```suggestion
		var newSearchPageItemSelector = '.SearchPage-Item';
		var newSearchPageAttachmentSelector = '.SearchPage-Attachment';
```

edalex-ian commented on Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:
```diff
@@ -329,9 +341,17 @@ export default function SearchResult({
+            key={id}
+            id={id}
+            button
+            className={`${classes.nested} NewSearchPage-Attachment`} // Give a class so each attachment can be dropped to the course list.
```
To keep things tight.... Should we attempt to use the class directly from `CourseList` - once it's added to the `declare` etc?

edalex-ian commented on Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:
```diff
@@ -452,7 +472,19 @@ export default function SearchResult({
+        id={uuid}
+        container
+        alignItems="center"
+        className="NewSearchPage-Item" // Give a class so each item can be dropped to the course list.
```
As above:

> To keep things tight.... Should we attempt to use the class directly from `CourseList` - once it's added to the `declare` etc?

edalex-ian reviewed (approved):
First up, fair effort @PenghaiZhang . Great wrangling of dragging and dropping across the boundaries of New UI to Legacy UI. :+1: 

Did you go back and confirm drag drop is all still cool in Legacy mode?

Other than that, just a few minor comments and queries. Was a bit tricky to review `courselist.js` with the reformatting and restructuring. But from what I can figure the algorithm and design is still pretty well the same, was just structural with support for the new classes for components from the React side.

So feel free to address what you like, and then merge on green. Let's get this puppy in there. :wink: :+1:

SammyIsConfused commented on Source/Plugins/Core/com.equella.core/js/tsrc/search/SearchPage.tsx:
```diff
@@ -182,6 +185,16 @@ const SearchPage = ({ updateTemplate }: TemplateUpdateProps) => {
+  useEffect(() => {
+    if (inSelectionSession) {
+      pagedSearchResult.results.forEach(({ uuid }) => {
+        prepareDraggable(uuid);
```
I have a sneaky suspicion that eslint will throw an `exhaustive-deps` warning for useEffects like these - since you refer to `inSelectionSession` and `prepareDraggable` without them being in the dependencies array. :thinking: Though it is likely that adding either of them will change functionality - probably in a way that doesn't work out. 

I don't have a recommendation on how to avoid that - just noting that it adds to the work to be addressed on squashing those warnings.

PenghaiZhang replied:
Yes I see this warning ` React Hook useEffect has a missing dependency: 'inSelectionSession'. Either include it or remove the dependency array `.

But since `inSelectionSession` is not in state, I can't see why we need to include it in the dependency array.

PenghaiZhang replied:
I can move the declaration of `inSelectionSession` inside the useEffect to avoid this warning though.

SammyIsConfused replied:
Yup, cool. So long as that all works it'll be happy.

SammyIsConfused commented on Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:
```diff
@@ -218,6 +221,15 @@ export default function SearchResult({
+  useEffect(() => {
+    if (inSelectionSession) {
+      attachmentsWithViewerDetails.forEach(({ attachment }) => {
+        prepareDraggable(attachment.id, false);
```
Same as above.

SammyIsConfused commented on Source/Plugins/Core/com.equella.core/resources/web/scripts/courselist.js:
```diff
@@ -63,115 +63,148 @@ var CourseList = {
+          folder);
+      }
+    }
+		if(doItems) {
```
Indenting looks pretty wacky here - I could be wrong.

SammyIsConfused replied:
Actually scratch that, it does line up. The nesting above threw me off.

SammyIsConfused commented on Source/Plugins/Core/com.equella.core/resources/web/scripts/courselist.js:
```diff
@@ -63,115 +63,148 @@ var CourseList = {
-		$(document).bind('equella_courselistupdate.courselist', courseListWatcher);
+    CourseList.prepareDroppable(
+      itemResultSelector +
+      ', ' + attachmentSelector +
```
Wondering if there is a more concise way to construct this string - but feel free to ignore.

PenghaiZhang replied:
We can put these strings into an array and then use `join()`.

edalex-ian replied:
I'm happy for this to stay as is for no other reason then to minimise any impact on the good ol' days.

PenghaiZhang replied:
Ah do you want me to revert 46cbf5ffc9eefe44b3d040c68f96e28fc154cdd9 ?

edalex-ian replied:
Up to you. Just make sure you check the "good ol' days". :wink:

SammyIsConfused reviewed (approved):
Looks good, nice work Penghai! :+1:

edalex-ian commented on Source/Plugins/Core/com.equella.core/resources/web/scripts/courselist.js:
```diff
@@ -97,23 +97,20 @@ var CourseList = {
-      dropCallback, clickFolderCallback)
+    var droppableSelectors =
+      [itemResultSelector, attachmentSelector, itemSummarySelector,
+        newSearchPageItemSelector, newSearchPageAttachmentSelector].join();
```
Just note that now you've gone from a separator of comma and space, to just comma. So as long as that space was to just make things pretty, then that's fine.

PenghaiZhang replied:
Yeah I have confirmed that removing the space does not break anything.
